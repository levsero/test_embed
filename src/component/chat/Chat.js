import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import _ from 'lodash';

import { ChatBox } from 'component/chat/ChatBox';
import { ChatHeader } from 'component/chat/ChatHeader';
import { ChatMessage } from 'component/chat/ChatMessage';
import { Container } from 'component/container/Container';
import { ScrollContainer } from 'component/container/ScrollContainer';
import { i18n } from 'service/i18n';
import { isMobileBrowser } from 'utility/devices';
import { endChat,
         sendMsg,
         setVisitorInfo,
         updateCurrentMsg,
         sendChatRating } from 'src/redux/modules/chat';

import { locals as styles } from './Chat.sass';

const mapStateToProps = (state) => {
  return { chat: state.chat };
};

class Chat extends Component {
  static propTypes = {
    chat: PropTypes.object.isRequired,
    endChat: PropTypes.func.isRequired,
    position: PropTypes.string,
    sendMsg: PropTypes.func.isRequired,
    setVisitorInfo: PropTypes.func.isRequired,
    style: PropTypes.object,
    updateCurrentMsg: PropTypes.func.isRequired,
    updateFrameSize: PropTypes.func,
    sendChatRating: PropTypes.func.isRequired
  };

  static defaultProps = {
    position: 'right',
    style: null,
    updateFrameSize: () => {}
  };

  constructor(props) {
    super(props);

    // Guard against WebWidget from accessing random
    // state attributes when state is not defined
    this.state = {};
  }

  updateUser = (user) => {
    this.props.setVisitorInfo({
      display_name: user.name || '', // eslint-disable-line camelcase
      email: user.email || ''
    });
  }

  renderChatLog = () => {
    const { chats } = this.props.chat;

    if (chats.size <= 0) return;

    const chatMessage = (data, key) => {
      return (<ChatMessage key={key} name={data.display_name} message={data.msg} nick={data.nick} />);
    };

    return _.chain([...chats.values()])
            .filter((m) => m.type === 'chat.msg')
            .map(chatMessage)
            .value();
  }

  renderChatEnded = () => {
    if (this.props.chat.chats.size <= 0 || this.props.chat.is_chatting) return;

    return (
      <div className={styles.chatEnd}>
        {i18n.t('embeddable_framework.chat.ended.label', { fallback: 'Chat Ended' })}
      </div>
    );
  }

  renderChatBox = () => {
    const { chat, sendMsg, updateCurrentMsg } = this.props;

    return (
      <ChatBox
        currentMessage={chat.currentMessage}
        sendMsg={sendMsg}
        updateCurrentMsg={updateCurrentMsg} />
    );
  }

  renderChatHeader = () => {
    const { chat, sendChatRating, endChat } = this.props;

    return (
      <ChatHeader
        rating={chat.rating}
        updateRating={sendChatRating}
        agents={chat.agents}
        endChat={endChat} />
    );
  }

  containerClasses = () => {
    return isMobileBrowser()
           ? styles.containerMobile
           : styles.container;
  }

  render = () => {
    setTimeout(() => this.props.updateFrameSize(), 0);

    return (
      <Container
        style={this.props.style}
        position={this.props.position}>
        <ScrollContainer
          title={i18n.t('embeddable_framework.helpCenter.label.link.chat')}
          headerContent={this.renderChatHeader()}
          headerClasses={styles.header}
          containerClasses={this.containerClasses()}
          footerClasses={styles.footer}
          footerContent={this.renderChatBox()}>
          <div className={styles.messages}>
            {this.renderChatLog()}
            {this.renderChatEnded()}
          </div>
        </ScrollContainer>
      </Container>
    );
  }
}

const actionCreators = {
  sendMsg,
  updateCurrentMsg,
  endChat,
  setVisitorInfo,
  sendChatRating
};

export default connect(mapStateToProps, actionCreators, null, { withRef: true })(Chat);
