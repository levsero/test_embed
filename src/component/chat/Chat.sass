@import '../../styles/_vars.scss';

.scrollContainerMobile {
  height: calc(100% - #{150/$font-size}rem);
  max-height: #{480/$font-size}rem;
}

.messages {
  margin-top: #{70/$font-size}rem;
  margin-bottom: #{24/$font-size}rem;
}

.header {
  composes: u-marginBM from global;
}

.footer {
  padding-bottom: #{8/$font-size}rem !important;
}

.chatEnd {
  composes: u-textCenter from global;
}

.chatEndPopup {
  composes: u-posAbsolute from global;
  bottom: #{10/$font-size}rem;
}

.chatEndPopupDescription {
  composes: u-paddingVL u-paddingLL from global;
}

.agentTyping {
  composes: u-posAbsolute u-paddingTS from global;
  bottom: #{103/$font-size}rem;
  color: $zd-color-monsoon;
  width: #{301/$font-size}rem;
  background: linear-gradient(to bottom, rgba($zd-color-white, 0), rgba($zd-color-white, 0.8));
}

.loadingEllipses {
  composes: u-inlineBlock from global;
  padding-right: #{4/$font-size}rem;
}

.loadingEllipsesItem {
  background-color: $zd-color-monsoon !important;
}
