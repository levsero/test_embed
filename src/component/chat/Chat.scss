@import '../../styles/_vars.scss';

.scrollContainerMobile {
  display: flex;
  flex-direction: column-reverse;

  .agentTyping {
    bottom: #{48/$font-size}rem;
  }
}

.container {
  height: #{$chat-height/$font-size}rem;
}

.mobileContainer {
  height: 100% !important;
}

.scrollContainer {
  composes: container;

  display: flex;
  flex-direction: column;
}

.scrollContainerContent {
  height: 100%;
  max-height: none;
}

.scrollContainerMessagesContent {
  composes: scrollContainerContent;

  margin-top: #{58/$font-size}rem;
}

.header {
  composes: u-marginBM from global;
}

.footer {
  padding-top: #{21/$font-size}rem !important;
  padding-bottom: #{4/$font-size}rem !important;
}

.footerMobile {
  padding: 0 !important;
  border-top: $thin-border-size solid $zd-color-gainsboro;
  margin-top: #{12/$font-size}rem !important;
}

.chatEnd {
  composes: u-textCenter from global;
}

.chatEndPopupDescription {
  composes: u-paddingVL u-paddingLL u-paddingRL from global;
}

.agentTyping {
  composes: u-posAbsolute u-paddingTS from global;
  bottom: #{97/$font-size}rem;
  color: $zd-color-aluminum;
  width: #{301/$font-size}rem;
  background: linear-gradient(to bottom, rgba($zd-color-white, 0), rgba($zd-color-white, 0.8));
}

.loadingEllipses {
  composes: u-inlineBlock from global;
  padding-right: #{2/$font-size}rem;
}

.loadingEllipsesItem {
  background-color: $zd-color-iron !important;
  width: #{4/$font-size}rem;
  height: #{4/$font-size}rem;
  margin-right: #{1/$font-size}rem;
  margin-bottom: #{2/$font-size}rem;
}

.queuePosition {
  composes: u-textCenter from global;
  margin-top: #{5/$font-size}rem;
  margin-bottom: #{10/$font-size}rem;
  color: $zd-color-aluminum;
}

.agentListBackButton {
  width: 100%;
  max-width: unset;
}

.reconnectContainer {
  composes: u-flex u-posFill u-posAbsolute u-flexAlignItemsCenter u-flexJustifyCenter from global;
}
