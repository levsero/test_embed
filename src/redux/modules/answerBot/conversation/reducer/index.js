import { combineReducers } from 'redux';

import lastScroll from './last-scroll';
import lastScreenClosed from './last-screen-closed';

export default combineReducers({
  lastScroll,
  lastScreenClosed
});
