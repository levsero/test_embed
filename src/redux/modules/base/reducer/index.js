import { combineReducers } from 'redux';

import activeEmbed from './base-active-embed';
import embeds from './base-embeds';
import zopim from './zopim';

export default combineReducers({
  activeEmbed,
  embeds,
  zopim
});

