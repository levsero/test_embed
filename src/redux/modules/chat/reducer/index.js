import { combineReducers } from 'redux';

import accountSettings from './account-settings/';
import accountStatus from './chat-account-status';
import agents from './chat-agents';
import chats from './chat-chats';
import connection from './chat-connection';
import currentMessage from './chat-current-message';
import departments from './chat-departments';
import isChatting from './chat-is-chatting';
import visitor from './chat-visitor';
import rating from './chat-rating';
import notification from './chat-notification';
import screen from './chat-screen';
import userSettings from './user-settings/';
import emailTranscript from './chat-email-transcript';
import editContactDetails from './chat-edit-contact-details';
import formState from './form-state';
import queuePosition from './chat-queue-position';
import offlineMessage from './chat-offline-message';
import menuVisible from './chat-menu-visibility';
import agentJoined from './chat-agent-joined';

export default combineReducers({
  accountSettings,
  account_status: accountStatus,
  agents,
  chats,
  connection,
  currentMessage,
  departments,
  is_chatting: isChatting,
  visitor,
  rating,
  notification,
  screen,
  userSettings,
  emailTranscript,
  formState,
  queuePosition,
  editContactDetails,
  menuVisible,
  offlineMessage,
  agentJoined
});
