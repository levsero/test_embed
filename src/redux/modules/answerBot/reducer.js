import { combineReducers } from 'redux';

import messages from './conversation/reducer/messages';
import conversation from './conversation/reducer';
import sessions from './sessions/reducer/sessions';
import currentMessage from './root/reducer/current-message';
import currentSessionID from './root/reducer/current-session-id';
import currentArticle from './root/reducer/current-article';
import currentContextualArticle from './root/reducer/current-contextual-article';
import currentScreen from './root/reducer/current-screen';
import greeted from './root/reducer/greeted';
import contextualSearchFinished from './root/reducer/contextual-search-finished';
import initialFallbackSuggested from './root/reducer/initial-fallback-suggested';

export default combineReducers({
  messages,
  conversation,
  currentMessage,
  sessions,
  currentSessionID,
  currentArticle,
  currentScreen,
  greeted,
  initialFallbackSuggested,
  contextualSearchFinished,
  currentContextualArticle
});
