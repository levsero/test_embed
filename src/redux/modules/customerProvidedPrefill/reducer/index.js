import { combineReducers } from 'redux'
import acknowledged from './acknowledged'
import types from './types'
import { API_RESET_WIDGET } from 'src/redux/modules/base/base-action-types'

const reducers = combineReducers({
  acknowledged,
  types
})

const customerProvidedPrefill = (state = {}, action) => {
  switch (action.type) {
    case API_RESET_WIDGET:
      return reducers(undefined, action)
    default:
      return reducers(state, action)
  }
}

export default customerProvidedPrefill
