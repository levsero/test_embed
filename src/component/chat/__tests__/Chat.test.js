import { render } from 'react-testing-library';
import React from 'react';

import { createStore } from 'redux';
import { Provider } from 'react-redux';

import reducer from 'src/redux/modules/reducer';
import Chat from '../Chat';

const selectors = require('src/redux/modules/chat/chat-selectors');
const showOfflineChatMock = jest.fn();

selectors.getShowOfflineChat = showOfflineChatMock;

const renderChat = () => {
  const store = createStore(reducer);

  return render(
    <Provider store={store}>
      <Chat
        getFrameContentDocument={() => {}}
        updateChatBackButtonVisibility={() => {}}
      />
    </Provider>,
  );
};

describe('show offline chat is true', () => {
  let rendered;

  beforeEach(() => {
    showOfflineChatMock.mockReturnValue(true);
    rendered = renderChat();
  });

  it('renders the offline form', () => {
    expect(rendered.getByText('Sorry, we are not online at the moment'))
      .toBeInTheDocument();
  });
});

describe('show offline chat is false', () => {
  let rendered;

  beforeEach(() => {
    showOfflineChatMock.mockReturnValue(false);

    rendered = renderChat();
  });

  it('renders the chatting screen', () => {
    expect(rendered.getByText('Live Support'))
      .toBeInTheDocument();
  });
});
