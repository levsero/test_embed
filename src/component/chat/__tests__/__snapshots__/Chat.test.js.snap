// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat sdk has not connected renders the loading page 1`] = `
<div>
  <div
    class="sc-htpNat iybLJI"
  >
    <div
      class="sc-ifAKCX iEHOTz"
    >
      <div
        class="sc-bZQynM sc-gzVnrw bQuIzm"
      >
        <h1
          class="sc-EHOje cygdwO"
          data-testid="widget-title"
        />
        <button
          aria-label="Minimize"
          class="IconButton-x15t1e-0 sc-htoDjs index__c-btn--icon___1Yl4x hhbqQ Button__StyledButton-sc-12lsl5w-0 index__c-btn___3y2Qe index__c-btn--basic___38OO7 index__c-btn--muted___ze6U0 index__c-btn--sm___3Kuhv lcPfmb"
          data-garden-container-id="keyboardfocus"
          data-garden-container-version="0.2.6"
          data-garden-id="buttons.icon_button"
          data-garden-version="7.0.2"
          tabindex="0"
          type="button"
        >
          <svg
            class="index__c-btn__icon___3V0Rl"
            data-garden-id="buttons.icon"
            data-garden-version="7.0.2"
            data-testid="Icon--dash"
            realfilename="dash-fill.svg"
          />
        </button>
      </div>
    </div>
    <div
      class="sc-dnqmqq LbvIn"
      data-testid="scroll-container-content"
    >
      <div
        class="sc-kEYyzF epLron"
      >
        <div
          class="styled-elements__LoadingPlaceholder-sc-19dhio6-6 dWxKQn"
          font-size="4rem"
          role="progressbar"
        >
          Â 
        </div>
      </div>
    </div>
    <footer
      class="sc-iwsKbI ikJxeP"
    >
      <div
        class="sc-gZMcBi kZYhVs"
      >
        <a
          class="icon"
          data-testid="Icon--zendesk"
          href="https://www.zendesk.com/embeddables/?utm_source=webwidget&utm_medium=poweredbyzendesk&utm_campaign=image"
          tabindex="-1"
          target="_blank"
        >
          <svg
            realfilename="widget-icon_zendesk.svg"
            title="zendesk"
          />
        </a>
      </div>
    </footer>
  </div>
</div>
`;
